File: DuDoong-Batch/src/main/java/band/gosrock/job/EventExpiration.java
package band.gosrock.job;


import band.gosrock.domain.domains.event.domain.Event;
import band.gosrock.domain.domains.event.service.EventService;
import java.time.LocalDateTime;
import java.util.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.batch.core.Job;
import org.springframework.batch.core.Step;
import org.springframework.batch.core.configuration.annotation.JobBuilderFactory;
import org.springframework.batch.core.configuration.annotation.JobScope;
import org.springframework.batch.core.configuration.annotation.StepBuilderFactory;
import org.springframework.batch.repeat.RepeatStatus;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Slf4j
@RequiredArgsConstructor
@Configuration
public class EventExpiration {

    private static final String JOB_NAME = "이벤트_자동만료";
    private static final String BEAN_PREFIX = JOB_NAME + "_";

    private final JobBuilderFactory jobBuilderFactory;
    private final StepBuilderFactory stepBuilderFactory;
    private final EventService eventService;

    //    @Bean(BEAN_PREFIX + "eventJobParameter")
    //    @JobScope
    //    public EventJobParameter eventJobParameter() {
    //        return new EventJobParameter(eventAdaptor);
    //    }
    //
    //    @Qualifier(BEAN_PREFIX + "eventJobParameter")
    //    private final EventJobParameter eventJobParameter;

    @Bean(JOB_NAME)
    public Job eventExpirationJob() {
        return jobBuilderFactory
                .get(JOB_NAME)
                .preventRestart()
                .start(eventExpirationStep())
                .build();
    }

    @Bean(BEAN_PREFIX + "step")
    @JobScope
    public Step eventExpirationStep() {
        return stepBuilderFactory
                .get(BEAN_PREFIX + "step")
                .tasklet(
                        (contribution, chunkContext) -> {
                            log.info(">>>>> 이벤트 자동 만료 작업 실행");
                            LocalDateTime now = LocalDateTime.now();

                            System.out.println("now = " + now);
                            List<Event> events = eventService.closeExpiredEventsEndAtBefore(now);
                            System.out.println("events.size() = " + events.size());

                            events.forEach(
                                    event -> System.out.println("closed event = " + event.getId()));

                            return RepeatStatus.FINISHED;
                        })
                .build();
    }
}

File: DuDoong-Batch/src/main/kotlin/band/gosrock/job/EventExpiration.kt
package band.gosrock.job

import band.gosrock.domain.domains.event.service.EventService
import band.gosrock.helper.slack.SlackEventExpirationSender
import band.gosrock.parameter.DateTimeJobParameter
import org.slf4j.LoggerFactory
import org.springframework.batch.core.Job
import org.springframework.batch.core.Step
import org.springframework.batch.core.configuration.annotation.JobBuilderFactory
import org.springframework.batch.core.configuration.annotation.JobScope
import org.springframework.batch.core.configuration.annotation.StepBuilderFactory
import org.springframework.batch.repeat.RepeatStatus
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class EventExpiration(
    private val jobBuilderFactory: JobBuilderFactory,
    private val stepBuilderFactory: StepBuilderFactory,
    private val slackEventExpirationSender: SlackEventExpirationSender,
    private val eventService: EventService,
) {
    private val log = LoggerFactory.getLogger(EventExpiration::class.java)

    companion object {
        private const val JOB_NAME = "이벤트_자동만료"
        private const val BEAN_PREFIX = "${JOB_NAME}_"
    }

    @Bean(BEAN_PREFIX + "dateTimeJobParameter")
    @JobScope
    fun dateTimeJobParameter(): DateTimeJobParameter = DateTimeJobParameter()

    @Bean(JOB_NAME)
    fun eventExpirationJob(): Job =
        jobBuilderFactory
            .get(JOB_NAME)
            .preventRestart()
            .start(eventExpirationStep())
            .build()

    @Bean(BEAN_PREFIX + "step")
    @JobScope
    fun eventExpirationStep(): Step =
        stepBuilderFactory
            .get(BEAN_PREFIX + "step")
            .tasklet { _, _ ->
                log.info(">>>>> 이벤트 자동 만료 작업 실행")
                val time = dateTimeJobParameter().getTime()
                val events = eventService.closeExpiredEventsEndAtBefore(time)
                slackEventExpirationSender.execute(time, events)
                RepeatStatus.FINISHED
            }
            .build()
}
